### DISCUSSION TITLE
Does MicroKernelTrait not properly support .php config files?

### DISCUSSION THREAD

**Discussion by chess-server** [symfony/symfony#46774] [Q&A]
I am planning to update my application to Symfony 6.0, following the SymfonyCast video. The application is already on 5.4, all deprecations are removed except one for the RouterConfiguration. 

As a first step, I have installed symfony/runtime and clearing my Kernel.php file as shown the video. This breaks my application. Deeper checks show that MicroKernelTrait does not fully support .php configuration files which I am using. 

In configureContainer, it loads only service.php from config, but would load only .yaml files from config/packages:

    private function configureContainer(ContainerConfigurator $container, LoaderInterface $loader, ContainerBuilder $builder): void
    {
        $configDir = $this->getConfigDir();

        $container->import($configDir.'/{packages}/*.yaml');
        $container->import($configDir.'/{packages}/'.$this->environment.'/*.yaml');

        if (is_file($configDir.'/services.yaml')) {
            $container->import($configDir.'/services.yaml');
            $container->import($configDir.'/{services}_'.$this->environment.'.yaml');
        } else {
            $container->import($configDir.'/{services}.php');
        }
    }

The same is valid for configureRoutes. 

I can overcome the issue by replacing the functions with using a constant like in the previous Kernel.php version.

    private const CONFIG_EXTS = '.{php,xml,yaml,yml}';

    private function configureContainer(ContainerConfigurator $container, LoaderInterface $loader, ContainerBuilder $builder): void
    {
        $configDir = $this->getConfigDir();

            $container->import($configDir.'/{packages}/*'.self::CONFIG_EXTS);
            $container->import($configDir.'/{packages}/'.$this->environment.'/*'.self::CONFIG_EXTS);

            $container->import($configDir.'/services'.self::CONFIG_EXTS);
            $container->import($configDir.'/{services}_'.$this->environment.self::CONFIG_EXTS);

    }

Is this a bug or is symfony expecting that services.php/routes.php loads all the php-files from config/packages and config/routes, resp.?

**Comment by ahundiak** [comment]
In 6.1 the trait has:
```
       $container->import($configDir.'/{packages}/*.{php,yaml}');
       $container->import($configDir.'/{packages}/'.$this->environment.'/*.{php,yaml}');
```
So php config/route files will work out of the box.  Not sure why 5.4/6.0 does not do this.  It's a bit annoying the way Kernel was cleaned out.  I used to tweak it all the time.

There are a couple of ways you could go here.  If your existing Kernel is working fine then I would just keep it.  At least until 6.1.

Or you could copy the code from the 6.1 trait to your own Kernel class and hopefully everything will work.

https://github.com/symfony/symfony/commit/3c76e962ccd84161e9b6517d5187061ec3bef539

**Reply by chess-server** [reply]
Thanks for your comment, my solution is basically what 6.1 is doing  (though 6.1 seems not to support xml-configurations anylonger).

**Reply by ahundiak** [reply]
Are you saying the xml configuration does not work at all or just that xml files are not picked up automatically?  I would think that
```
$container->import($configDir.'/{packages}/*.{php,yaml,xml}');
```
would work as expected though having three formats in one app might start to be a bit confusing.

**Reply by chess-server** [reply]
From a simple look at your code fragment

       $container->import($configDir.'/{packages}/*.{php,yaml}');
       $container->import($configDir.'/{packages}/'.$this->environment.'/*.{php,yaml}');

I would say it does pick only php and yaml files. It would not even load applications which are configured through XML **only**. The code simple does not check for XML. Symfony 5.4/6.0 load only yaml-configs completely. It does load a services.php file if services.yaml does not exist, but even then it would not load php files from config/packages. And it does not load any xml-files. Adding xml into the curly brackets solves the problem. But new users might get into trouble. It took me a day to find out what was wrong.

**Reply by ahundiak** [reply]
Why all config files are not automatically loaded regardless of their extension is one of those mysteries that date back to the dawn of time.  Or at least to the release of Symfony 2.0.  Some things we are just not meant to know.

I have seen many similar questions most of which were resolved but pointing developers to src/Kernel.php and the configure methods.  It's knowing where to look that is difficult the first time around.
